pkg/api/router.go

```diff
@@ -15,49 +16,81 @@ import (
+		writeHandler = authWrapper(withWarnLog("trying to send metrics to write API while connector is in read-only mode", http.NotFoundHandler()))
 	}
 
 	router.Post("/write", writeHandler)
 
-	readHandler := timeHandler(metrics.HTTPRequestDuration, "read", Read(client, metrics))
+	readHandler := authWrapper(timeHandler(metrics.HTTPRequestDuration, "read", Read(client, metrics)))
 	router.Get("/read", readHandler)
 	router.Post("/read", readHandler)
 
-	deleteHandler := timeHandler(metrics.HTTPRequestDuration, "delete_series", Delete(apiConf, client))
+	deleteHandler := authWrapper(timeHandler(metrics.HTTPRequestDuration, "delete_series", Delete(apiConf, client)))
 	router.Put("/delete_series", deleteHandler)
 	router.Post("/delete_series", deleteHandler)
 
 	queryable := client.GetQueryable()
 	queryEngine := query.NewEngine(log.GetLogger(), time.Minute)
-	queryHandler := timeHandler(metrics.HTTPRequestDuration, "query", Query(apiConf, queryEngine, queryable))
+	queryHandler := authWrapper(timeHandler(metrics.HTTPRequestDuration, "query", Query(apiConf, queryEngine, queryable)))
 	router.Get("/api/v1/query", queryHandler)
 	router.Post("/api/v1/query", queryHandler)
 
-	queryRangeHandler := timeHandler(metrics.HTTPRequestDuration, "query_range", QueryRange(apiConf, queryEngine, queryable))
+	queryRangeHandler := authWrapper(timeHandler(metrics.HTTPRequestDuration, "query_range", QueryRange(apiConf, queryEngine, queryable)))
 	router.Get("/api/v1/query_range", queryRangeHandler)
 	router.Post("/api/v1/query_range", queryRangeHandler)
 
-	seriesHandler := timeHandler(metrics.HTTPRequestDuration, "series", Series(apiConf, queryable))
+	seriesHandler := authWrapper(timeHandler(metrics.HTTPRequestDuration, "series", Series(apiConf, queryable)))
 	router.Get("/api/v1/series", seriesHandler)
 	router.Post("/api/v1/series", seriesHandler)
 
-	labelsHandler := timeHandler(metrics.HTTPRequestDuration, "labels", Labels(apiConf, queryable))
+	labelsHandler := authWrapper(timeHandler(metrics.HTTPRequestDuration, "labels", Labels(apiConf, queryable)))
 	router.Get("/api/v1/labels", labelsHandler)
 	router.Post("/api/v1/labels", labelsHandler)
 
-	labelValuesHandler := timeHandler(metrics.HTTPRequestDuration, "label/:name/values", LabelValues(apiConf, queryable))
+	labelValuesHandler := authWrapper(timeHandler(metrics.HTTPRequestDuration, "label/:name/values", LabelValues(apiConf, queryable)))
 	router.Get("/api/v1/label/:name/values", labelValuesHandler)
 
 	router.Get("/healthz", Health(client))
 
 	return router
 }
 
+func authHandler(cfg *Config, handler http.HandlerFunc) http.HandlerFunc {
+	if cfg.BasicAuthUsername != "" {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			user, pass, ok := r.BasicAuth()
+			if !ok || cfg.BasicAuthUsername != user || cfg.BasicAuthPassword != pass {
+				log.Error("msg", "Unauthorized access to endpoint, invalid username or password")
```

Comment:
we probably want to limit the frequency with which these messages are printed and/or just print the first one. I do think we need to add a Prometheus metric to track the number of unauthorized accesses (counter)